wfsim/strax_interface.py

```diff
@@ -621,147 +480,285 @@ def _setup(self):
-        
-        #Rate in ns^-1
-        rate = self.config['event_rate']/wfsim.units.s
-        start = self.config['event_start']
-        stop = self.config['event_stop']
-        if stop == -1:
-            stop = self.instructions['g4id'][-1]
-
-        timings= np.random.uniform(start/rate,stop/rate,stop-start).astype(np.int64)
-        timings.sort()
 
-        #For tpc we have multiple instructions per g4id. For nveto we only have one
-        counter=Counter(self.instructions['g4id'])
-        i_per_id = [counter[k] for k in range(start,stop)]
-        timings_tpc=np.repeat(timings,i_per_id)
-        self.instructions['time']+=timings_tpc
+        # If no event_stop is specified set it to the maximum (-1=do all events)
+        if self.config['entry_stop'] is None:
+            self.config['entry_start'] = np.min(self.g4id)
+            self.config['entry_stop'] = np.max(self.g4id) + 1
+        log.debug('Entry stop set at %d, g4id min %d max %d'
+                  % (self.config['entry_stop'], np.min(self.g4id), np.max(self.g4id)))
+
+        # Convert rate from Hz to ns^-1
+        rate = self.config['event_rate'] / 1e9
+        # Add half interval to avoid time 0
+        timings = np.random.uniform((self.config['entry_start'] + 0.5) / rate, 
+                                    (self.config['entry_stop'] + 0.5) / rate, 
+                                    self.config['entry_stop'] - self.config['entry_start'])
+        timings = np.sort(timings).astype(np.int64)
+        max_time = int((self.config['entry_stop'] + 0.5) / rate)
+
+        # For tpc we have multiple instructions per g4id.
+        if 'tpc' in self.config['targets']:
+            i_timings = np.searchsorted(np.arange(self.config['entry_start'], self.config['entry_stop']),
+                                        self.instructions_epix['g4id'])
+            self.instructions_epix['time'] += timings[i_timings]
+
+            extra_long = self.instructions_epix['time'] > max_time
+            self.instructions_epix = self.instructions_epix[~extra_long]
+            log.warning('Found and removing %d epix instructions later than maximum time %d'
+                        % (extra_long.sum(), max_time))
+            self.instrutions_epix = np.sort(self.instructions_epix, order='time')
+
+        # nveto instruction doesn't carry physical time delay, so the time is overwritten
+        if 'nveto' in self.config['targets']:
+            i_timings = np.searchsorted(np.arange(self.config['entry_start'], self.config['entry_stop']),
+                                        self.instructions_nveto['g4id'])
+            self.instructions_nveto['time'] += timings[i_timings]
+
+            extra_long = self.instructions_nveto['time'] > max_time
+            self.instructions_nveto = self.instructions_nveto[~extra_long]
+            log.warning('Found and removing %d nveto instructions later than maximum time %d'
+                        % (extra_long.sum(), max_time))
```

Comment:
Here, and in the TPC case a few lines above, the `max_time` reference is the G4 event zero time. We need to discuss this with actual examples, as valid events can arise that simply take longer to produce the first interaction wrt this zero time.

Reply:
Both the `self.instructions_nveto['time']` and `max_time` are absolute time at this point. The instructions removed here are cases that the time of the instruction is so late that we can not fit them into this batch, (max time defined by `entry_stop / event_rate`

Reply:
Thanks @zhut19, now I see `max_time = int((self.config['entry_stop'] + 0.5) / rate)` a few lines above. Two comments then:

1. With that definition, doesn't `max_time` risk to be identical than the starting time of the last entry? Shouldn't we add a buffer to `max_time`?
2. We still need to discuss what to do with the late G4 entries (we apply a cut in epix to entries 10ms after the first one in in LXe (so not in absolute event time), but the same isn't done in the nVeto side.

Reply:
1. 
    - Depending on our event rate, even with a typical delay between the signal and the G4 time of 1ms, and an event rate of 100 Hz, we would have 0.1 events outside the max time per batch.
    -  Adding a buffer after max_time doesn't solve the issue completely since there is always this delay (between the signal and the G4 time), but can very likely create overlapping with the next batch.
2.
    - The logic implemented here is a bit different, thus I am not sure any 10 ms cut-off is necessary. 
    - Here, for example, uniform timing is sampled for each G4id and added to all signal timings to make this combined timing, we only remove the really delayed signals that happen after the max time. In which case for the first entry, the delay time cut-off is ~ 10 seconds (100 Hz, 1000 evt/batch), but for the last entry, the delay time cut-off is ~ 10 milliseconds (100 Hz).
    - For nveto, the newly added function in https://github.com/XENONnT/WFSim/pull/125/commits/99572a8c4b7845b021c643846b08a5db6cb59c0a ensures that each instruction time is actually the signal time with the photons of each instruction happen at the latest 1ms afterward. And in the set timing step for nveto, the same time combining and max time cut-off logic is applied.

Reply:
>     1. * Depending on our event rate, even with a typical delay between the signal and the G4 time of 1ms, and an event rate of 100 Hz, we would have 0.1 events outside the max time per batch.
>        * Adding a buffer after max_time doesn't solve the issue completely since there is always this delay (between the signal and the G4 time), but can very likely create overlapping with the next batch.
>

Just to confirm, here you say "next batch" as referring to future stages in which we merge various WFSim outputs into superruns, correct?
 
>     2. * The logic implemented here is a bit different, thus I am not sure any 10 ms cut-off is necessary.
>        * Here, for example, uniform timing is sampled for each G4id and added to all signal timings to make this combined timing, we only remove the really delayed signals that happen after the max time. In which case for the first entry, the delay time cut-off is ~ 10 seconds (100 Hz, 1000 evt/batch), but for the last entry, the delay time cut-off is ~ 10 milliseconds (100 Hz).

Okay, in this case I will make sure to test a few different type of simulations and see whether we remove lots of epix delayed instructions (there's now a printout for this thanks to you!). But I agree that the epix cutoff doesn't need to be matched with nVeto a priori (will keep thinking about this...).

>        * For nveto, the newly added function in [99572a8](https://github.com/XENONnT/WFSim/commit/99572a8c4b7845b021c643846b08a5db6cb59c0a) ensures that each instruction time is actually the signal time with the photons of each instruction happen at the latest 1ms afterward. And in the set timing step for nveto, the same time combining and max time cut-off logic is applied.

👍